docs/book/v3/features/helpers/template-variable-container.md

```diff
@@ -0,0 +1,149 @@
+# Template Variable Container
+
+> - Since zend-expressive-helpers 5.3.0
+
+[zend-expressive-template](../template/intro.md) provides the method
+[Zend\Expressive\Template\TemplateRendererInterface::addDefaultParam()](../template/interface.md#default-params)
+for providing template variables that should be available to any template.
+
+One common use case for this is to set things such as the current user, current
+section of the website, currently matched route, etc. Unfortunately, because the
+method changes the internal state of the renderer, this can cause problems in an
+async environment, such as [Swoole](https://docs.zendframework.com/zend-expressive-swoole), 
+where those changes will persist for parallel and subsequent requests.
+
+To provide a stateless alternative, you can create a `Zend\Expressive\Helper\Template\TemplateVariableContainer`
```

Comment:
I think this is a very important information: _alternative_ – this should be clarified and highlighted maybe with a block-quote. When should be used `addDefaultParam` and when `TemplateVariableContainer`.

Reply:
The "when" part is really up for debate.

It **MUST** be used when in an async environment, and you need to add pipeline-specific default parameters.

In non-async environments, you could still use it, as it makes for a more explicit approach that, again, avoids state side effects.

I'll see if I can draft something reasonable.